File: engine/julia/tactical_options_discipline_engine.jl
# Module AB5 (Julia): Tactical Options Structuring & Execution Discipline Engine
# High-precision options payoff modeling, Iron Condor wing insurance analysis, and capital sizing rules.

module TacticalOptionsDisciplineEngine

export calculate_position_size, structure_iron_condor

function calculate_position_size(equity::Float64, entry::Float64, stop::Float64, is_aggressive::Bool=false)
    risk_frac = is_aggressive ? 0.02 : 0.01
    dollar_risk = equity * risk_frac
    per_share_risk = abs(entry - stop)
    shares = per_share_risk > 0.0 ? floor(Int, dollar_risk / per_share_risk) : 0
    contracts = floor(Int, shares / 100)
    return (
        dollar_risk = round(dollar_risk, digits=2),
        shares = shares,
        contracts = contracts
    )
end

function structure_iron_condor(k1::Float64, k2::Float64, k3::Float64, k4::Float64,
                               p_short::Float64, p_long::Float64, c_short::Float64, c_long::Float64)
    put_credit = p_short - p_long
    call_credit = c_short - c_long
    total_credit = (put_credit + call_credit) * 100.0
    wing_width = (k2 - k1) * 100.0
    max_loss = max(0.0, wing_width - total_credit)
    rr = max_loss > 0.0 ? total_credit / max_loss : 0.0
    
    return (
        net_credit = round(total_credit, digits=2),
        wing_width = round(wing_width, digits=2),
        max_loss = round(max_loss, digits=2),
        reward_to_risk = round(rr, digits=2)
    )
end

end # module

File: engine/rust/src/tactical_options_discipline_engine.rs
//! Module AB4 (Rust): Tactical Options Structuring & Execution Discipline Engine.
//! High-speed options sizing, Iron Condor analyzer, and discipline risk enforcement.

#[repr(C, align(64))]
#[derive(Debug, Clone, Copy)]
pub struct TacticalOptionsDisciplineState {
    pub account_equity: f32,
    pub max_dollar_risk: f32,
    pub entry_price: f32,
    pub stop_loss_price: f32,
    pub take_profit_price: f32,
    pub iron_condor_net_credit: f32,
    pub iron_condor_max_loss: f32,
    pub iron_condor_rr: f32,
    pub recommended_shares: u32,
    pub recommended_contracts: u32,
    pub oco_active_status: u32,
    pub discipline_lock_flag: u32,
    pub _padding: [u8; 16],
}

pub struct TacticalOptionsDisciplineEngine;

impl TacticalOptionsDisciplineEngine {
    pub fn new_state() -> TacticalOptionsDisciplineState {
        TacticalOptionsDisciplineState {
            account_equity: 10000.0,
            max_dollar_risk: 100.0,
            entry_price: 0.0,
            stop_loss_price: 0.0,
            take_profit_price: 0.0,
            iron_condor_net_credit: 0.0,
            iron_condor_max_loss: 0.0,
            iron_condor_rr: 0.0,
            recommended_shares: 0,
            recommended_contracts: 0,
            oco_active_status: 0,
            discipline_lock_flag: 0,
            _padding: [0; 16],
        }
    }

    pub fn compute_sizing_and_condor(
        state: &mut TacticalOptionsDisciplineState,
        equity: f32,
        entry: f32,
        stop: f32,
        tp: f32,
        k1: f32, k2: f32, _k3: f32, _k4: f32,
        p_short: f32, p_long: f32, c_short: f32, c_long: f32,
    ) {
        state.account_equity = equity;
        state.entry_price = entry;
        state.stop_loss_price = stop;
        state.take_profit_price = tp;

        let risk_fraction = 0.01;
        state.max_dollar_risk = equity * risk_fraction;
        let per_share_risk = (entry - stop).abs();
        state.recommended_shares = if per_share_risk > 0.0 { (state.max_dollar_risk / per_share_risk).floor() as u32 } else { 0 };
        state.recommended_contracts = state.recommended_shares / 100;

        let put_credit = p_short - p_long;
        let call_credit = c_short - c_long;
        state.iron_condor_net_credit = (put_credit + call_credit) * 100.0;
        let wing_width = (k2 - k1) * 100.0;
        state.iron_condor_max_loss = (wing_width - state.iron_condor_net_credit).max(0.0);
        state.iron_condor_rr = if state.iron_condor_max_loss > 0.0 { state.iron_condor_net_credit / state.iron_condor_max_loss } else { 0.0 };

        state.oco_active_status = 1;
        state.discipline_lock_flag = 0;
    }
}
